diff --git a/ChangeLog b/ChangeLog
index 634d2a1..8d2a201 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,14 @@
+2012-12-26  KwangYong Choi  <ky0.choi@samsung.com>
+
+        [EFL] Enable MHTML feature
+        https://bugs.webkit.org/show_bug.cgi?id=xxxxx
+
+        Reviewed by NOBODY (OOPS!).
+
+        ENABLE_MHTML is now ON for EFL.
+
+        * Source/cmake/OptionsEfl.cmake:
+
 2012-12-24  Kondapally Kalyan  <kalyan.kondapally@intel.com>
 
         [EFL [WebGL] GLES2 detection is broken.
diff --git a/Source/WebCore/ChangeLog b/Source/WebCore/ChangeLog
index 046ef14..4ac5387 100644
--- a/Source/WebCore/ChangeLog
+++ b/Source/WebCore/ChangeLog
@@ -1,3 +1,20 @@
+2012-12-26  KwangYong Choi  <ky0.choi@samsung.com>
+
+        [EFL] Enable MHTML feature
+        https://bugs.webkit.org/show_bug.cgi?id=xxxxx
+
+        Reviewed by NOBODY (OOPS!).
+
+        Add application/x-mimearchive type to archive mime types.
+        Add mht and mhtml to extension map.
+
+        No new tests. Existing layout tests pass.
+
+        * loader/archive/ArchiveFactory.cpp:
+        (WebCore::archiveMIMETypes):
+        * platform/efl/MIMETypeRegistryEfl.cpp:
+        (WebCore):
+
 2012-12-26  Huang Dongsung  <luxtella@company100.net>
 
         [TexMap] A Minor optimization of GraphicsLayerTransform.
diff --git a/Source/WebCore/loader/archive/ArchiveFactory.cpp b/Source/WebCore/loader/archive/ArchiveFactory.cpp
index 4ebb52b..a989154 100644
--- a/Source/WebCore/loader/archive/ArchiveFactory.cpp
+++ b/Source/WebCore/loader/archive/ArchiveFactory.cpp
@@ -70,7 +70,7 @@ static ArchiveMIMETypesMap& archiveMIMETypes()
     mimeTypes.set("multipart/related", archiveFactoryCreate<MHTMLArchive>);
 #if PLATFORM(GTK)
     mimeTypes.set("message/rfc822", archiveFactoryCreate<MHTMLArchive>);
-#elif PLATFORM(QT)
+#elif PLATFORM(QT) || PLATFORM(EFL)
     mimeTypes.set("application/x-mimearchive", archiveFactoryCreate<MHTMLArchive>);
 #endif
 #endif
diff --git a/Source/WebCore/platform/efl/MIMETypeRegistryEfl.cpp b/Source/WebCore/platform/efl/MIMETypeRegistryEfl.cpp
index d883e5d..540bb2d 100644
--- a/Source/WebCore/platform/efl/MIMETypeRegistryEfl.cpp
+++ b/Source/WebCore/platform/efl/MIMETypeRegistryEfl.cpp
@@ -51,6 +51,8 @@ static const ExtensionMap extensionMap[] = {
     { "jpeg", "image/jpeg" },
     { "jpg", "image/jpeg" },
     { "js", "application/x-javascript" },
+    { "mht", "application/x-mimearchive" },
+    { "mhtml", "application/x-mimearchive" },
     { "mng", "video/x-mng" },
     { "pbm", "image/x-portable-bitmap" },
     { "pgm", "image/x-portable-graymap" },
diff --git a/Source/WebKit2/CMakeLists.txt b/Source/WebKit2/CMakeLists.txt
index 95f9acb..ca1902f 100644
--- a/Source/WebKit2/CMakeLists.txt
+++ b/Source/WebKit2/CMakeLists.txt
@@ -69,6 +69,7 @@ set(WebKit2_INCLUDE_DIRECTORIES
     "${WEBCORE_DIR}/html/track"
     "${WEBCORE_DIR}/inspector"
     "${WEBCORE_DIR}/loader"
+    "${WEBCORE_DIR}/loader/archive"
     "${WEBCORE_DIR}/loader/icon"
     "${WEBCORE_DIR}/loader/cache"
     "${WEBCORE_DIR}/page"
diff --git a/Source/WebKit2/ChangeLog b/Source/WebKit2/ChangeLog
index f1f6d4c..7709abc 100644
--- a/Source/WebKit2/ChangeLog
+++ b/Source/WebKit2/ChangeLog
@@ -1,3 +1,14 @@
+2012-12-26  KwangYong Choi  <ky0.choi@samsung.com>
+
+        [EFL] Enable MHTML feature
+        https://bugs.webkit.org/show_bug.cgi?id=xxxxx
+
+        Reviewed by NOBODY (OOPS!).
+
+        Add missing include directory for ENABLE_MHTML.
+
+        * CMakeLists.txt:
+
 2012-12-26  Csaba Osztrogonác  <ossy@webkit.org>
 
         Unreviewed Xmas buildfix after r138479.
diff --git a/Source/cmake/OptionsEfl.cmake b/Source/cmake/OptionsEfl.cmake
index 800a709..1fc3def 100644
--- a/Source/cmake/OptionsEfl.cmake
+++ b/Source/cmake/OptionsEfl.cmake
@@ -70,6 +70,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_LINK_PREFETCH ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_LLINT ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MEDIA_CAPTURE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MEMORY_SAMPLER ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MHTML ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MICRODATA ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_NAVIGATOR_CONTENT_UTILS ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_NETSCAPE_PLUGIN_API ON)
diff --git a/Tools/ChangeLog b/Tools/ChangeLog
index 7f40db6..5890dbf 100644
--- a/Tools/ChangeLog
+++ b/Tools/ChangeLog
@@ -1,3 +1,14 @@
+2012-12-26  KwangYong Choi  <ky0.choi@samsung.com>
+
+        [EFL] Enable MHTML feature
+        https://bugs.webkit.org/show_bug.cgi?id=xxxxx
+
+        Reviewed by NOBODY (OOPS!).
+
+        Enable MHTML feature for EFL port by default.
+
+        * Scripts/webkitperl/FeatureList.pm:
+
 2012-12-25  Christophe Dumez  <christophe.dumez@intel.com>
 
         [EFL][WK2] Refactor snapshot taking code
diff --git a/Tools/Scripts/webkitperl/FeatureList.pm b/Tools/Scripts/webkitperl/FeatureList.pm
index 2eddb26..975f2bf 100644
--- a/Tools/Scripts/webkitperl/FeatureList.pm
+++ b/Tools/Scripts/webkitperl/FeatureList.pm
@@ -332,7 +332,7 @@ my @features = (
       define => "ENABLE_METER_ELEMENT", default => !isAppleWinWebKit(), value => \$meterTagSupport },
 
     { option => "mhtml", desc => "Toggle MHTML support",
-      define => "ENABLE_MHTML", default => isGtk(), value => \$mhtmlSupport },
+      define => "ENABLE_MHTML", default => (isGtk() || isEfl()), value => \$mhtmlSupport },
 
     { option => "microdata", desc => "Toggle Microdata support",
       define => "ENABLE_MICRODATA", default => (isEfl() || isBlackBerry() || isGtk()), value => \$microdataSupport },
