diff --git a/autogen.sh b/autogen.sh
index 43f499b..0b9c793 100755
--- a/autogen.sh
+++ b/autogen.sh
@@ -47,7 +47,7 @@ if test -d po ; then
   touch -t 200001010000 po/gst-plugins-bad-1.0.pot
 fi
 
-CONFIGURE_DEF_OPT='--enable-maintainer-mode --enable-gtk-doc'
+CONFIGURE_DEF_OPT='--enable-maintainer-mode'
 
 if test "x$package" = "xgstreamer"; then
   CONFIGURE_DEF_OPT="$CONFIGURE_DEF_OPT --enable-docbook --enable-failing-tests --enable-poisoning"
